src/System.Memory/src/System/Memory.cs

```diff
@@ -145,7 +144,9 @@ public Memory<T> Slice(int start)
 
             if (_index < 0)
                 return new Memory<T>((OwnedMemory<T>)_arrayOrOwnedMemory, (_index & RemoveOwnedFlagBitMask) + start, _length - start);
-            return new Memory<T>((T[])_arrayOrOwnedMemory, _index + start, _length - start);
+            if (_arrayOrOwnedMemory != null)
+                return new Memory<T>((T[])_arrayOrOwnedMemory, _index + start, _length - start);
```

Comment:
How does the JIT do at eliminating the null check in the inlined Memory ctor?  For that matter, how does it do at eliminating the range checks that also shouldn't be necessary?  I'm wondering if we should have a private ctor that let's us bypass those additional checks, especially since this is aggressively inlined.

Reply:
(same question in other cases of this)

Reply:
> How does the JIT do at eliminating the null check in the inlined Memory ctor?

It looks like the JIT eliminates the extra null check (testing on a sample app and confirming with Slice before/after). It doesn't eliminate the range checks though.

![image](https://user-images.githubusercontent.com/6527137/32476847-fb744fce-c32e-11e7-8715-176065318e48.png)

~For some reason though, comparing the assembly for Slice (before this change and after) still shows an extra check, in a net core 2.0 app~:
!~[image](https://user-images.githubusercontent.com/6527137/32476612-492f3974-c32d-11e7-85d3-1d09fc9c9883.png)~

> I'm wondering if we should have a private ctor that let's us bypass those additional checks, especially since this is aggressively inlined.

How would we write such a ctor overload? We can't change the ctor signature by renaming it or changing the arguments.

Reply:
This is pretty much the same idea as #24296 which @jkotas wasn't keen on. Looks like the null checks should get eliminated but the range checks won't.

Reply:
It doesn't matter at this point.  I already added such a private ctor in support of string, and you can see these checks are gone in https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Memory.cs#L164-L171, and being removed in corefx here: https://github.com/dotnet/corefx/pull/25120/files#diff-5c8b5ed1bac9b33228ea179f996dc240R152.

Reply:
This has been done in the CoreCLR version of the file already - I just have not propagated here yet:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Memory.cs#L152

The optimization here is internal implementation detail of the type, not a public API - it is what makes it different from #24296.